clang/lib/Sema/SemaFunctionEffects.cpp

```diff
@@ -0,0 +1,1572 @@
+  DeclDisallowsInference,
+
+  // These both apply to indirect calls. The difference is that sometimes
+  // we have an actual Decl (generally a variable) which is the function
+  // pointer being called, and sometimes, typically due to a cast, we only
+  // have an expression.
+  CallsDeclWithoutEffect,
+  CallsExprWithoutEffect,
+};
+
+// Information about the AST context in which a violation was found, so
+// that diagnostics can point to the correct source.
+class ViolationSite {
+public:
+  enum class Kind : uint8_t {
+    Default = 0, // Function body.
+    MemberInitializer = 1,
+    DefaultArgExpr = 2
+  };
+
+private:
+  llvm::PointerIntPair<CXXDefaultArgExpr *, 2, Kind> Impl;
+
+public:
+  ViolationSite() = default;
+
+  explicit ViolationSite(CXXDefaultArgExpr *E)
+      : Impl(E, Kind::DefaultArgExpr) {}
+
+  Kind kind() const { return static_cast<Kind>(Impl.getInt()); }
+  CXXDefaultArgExpr *defaultArgExpr() const { return Impl.getPointer(); }
+
+  void setKind(Kind K) { Impl.setPointerAndInt(nullptr, K); }
+};
+
+// Represents a violation of the rules, potentially for the entire duration of
+// the analysis phase, in order to refer to it when explaining why a caller has
+// been made unsafe by a callee. Can be transformed into either a Diagnostic
+// (warning or a note), depending on whether the violation pertains to a
+// function failing to be verifed as holding an effect vs. a function failing to
+// be inferred as holding that effect.
+struct Violation {
+  FunctionEffect Effect;
+  FunctionEffect
+      CalleeEffectPreventingInference; // Only for certain IDs; can be None.
+  ViolationID ID = ViolationID::None;
+  ViolationSite Site;
+  SourceLocation Loc;
+  const Decl *Callee = nullptr; // Only valid for Calls*.
+
+  Violation() = default;
+
+  Violation(FunctionEffect Effect, ViolationID ID, ViolationSite VS,
+            SourceLocation Loc, const Decl *Callee = nullptr,
+            std::optional<FunctionEffect> CalleeEffect = std::nullopt)
+      : Effect(Effect), CalleeEffectPreventingInference(
+                            CalleeEffect.value_or(FunctionEffect())),
+        ID(ID), Site(VS), Loc(Loc), Callee(Callee) {}
+
+  unsigned diagnosticSelectIndex() const {
+    return unsigned(ID) - unsigned(ViolationID::BaseDiagnosticIndex);
+  }
+};
+
+enum class SpecialFuncType : uint8_t { None, OperatorNew, OperatorDelete };
+enum class CallableType : uint8_t {
+  // Unknown: probably function pointer.
+  Unknown,
+  Function,
+  Virtual,
+  Block
+};
+
+// Return whether a function's effects CAN be verified.
+// The question of whether it SHOULD be verified is independent.
+static bool functionIsVerifiable(const FunctionDecl *FD) {
+  if (FD->isTrivial()) {
+    // Otherwise `struct x { int a; };` would have an unverifiable default
+    // constructor.
+    return true;
+  }
+  return FD->hasBody();
+}
+
+static bool isNoexcept(const FunctionDecl *FD) {
+  const auto *FPT = FD->getType()->castAs<FunctionProtoType>();
```

Comment:
We do have tests for prototypeless functions, right?

Reply:
Yeah, in attr-nonblocking-sema.c:

```
// Tests for a few cases involving C functions without prototypes.

void noproto() __attribute__((nonblocking)) // expected-error {{'nonblocking' function must have a prototype}}
{
}

// This will succeed
void noproto(void) __attribute__((blocking));

// A redeclaration isn't any different - a prototype is required.
void f1(void);
void f1() __attribute__((nonblocking)); // expected-error {{'nonblocking' function must have a prototype}}
```

Reply:
Ah, but do we also have tests that perform effect analysis on C code? Just trying to make sure that that `castAs` (and maybe other places where we require a `FunctionProtoType`) doesn’t assert on us—or is this guarded by a `getLangOpts().CPlusPlus` check that I missed?

Reply:
Thanks, this exposed that isNoexcept() was the one place that wasn't prepared for a prototype-less function type. For long-term safety I fixed that, but I also short-circuited the place that was calling it. In plain C, a terminating call like `abort()` can be detected and excluded from analysis with just `noreturn`, since `noexcept` isn't possible, and just too bad if it calls `longjmp()` (?).

Reply:
`longjmp()` should be detected as a builtin I think, so that should solve that.

Reply:
Other than that, noreturn = probably terminates stgm in C.

Reply:
I hope I haven't misunderstood you here. I did add a test involving a `noreturn` function in attr-nonblocking-constraints.c (distilled):

```
// Has no prototype, noreturn.
[[noreturn]]
void aborts();

void nb2(void) __attribute__((nonblocking)) {
  aborts(); // no diagnostic because it's noreturn.
}
```

The idea is that abort/terminate is a "force majeure", outside the scope of the effect analysis.

Reply:
Yeah, no that one is fine, but builtin functions should still overrule that, e.g. if `longjmp` is declared `noreturn`, it should be blocking regardless because it’s a builtin function we consider to be blocking, so what I was saying is that we should have a test for `longjmp` specifically (or any other builtin function that can conceivably be declared `noreturn` but blocks).

Reply:
Ah gotcha. I looked through all the `noreturn` builtins and most seem like they do trap/abort/exit. I don't know if a typical implementation `longjmp()` really does allocate anything on the heap or block, but philosophically I agree that it's like throwing an exception, which does allocate.

Reply:
> a typical implementation of `longjmp()` really does allocate anything on the heap or block, but philosophically I agree that it's like throwing an exception, which does allocate.

Not just philosophically; some implementations literally just throw an exception for `longjmp()`.

Reply:
(or at least some platforms—e.g. Windows, I think—unwind the stack when `longjmp()` is called; I don’t know if it allocates tho, I’m not really a backend guy)